### PR TITLE
Migrate the Toolbar to React

### PR DESCRIPTION
**Pull request by etrobert**
The graph is currently still a big chunk of js

It used to get wether we're in link mode or not by looking at a global checkbox. Kept a similar one here to not break the graph

The graph sends `selectionchanged` events whenever the selected tasks in the graph changes (Ctrl-A, click on background, ...)

We update `tasksSelected` with the `useTasksSelected` hook accordingly

Note: The `App.tsx` is starting to get fat 🤔

### CODE REVIEW

File: src/scripts/App.tsx [hunk 1]
```diff
@@ -42,12 +51,52 @@ const App = (): JSX.Element => {
...
+  const tasksSelected = useTasksSelected();
+
+  const renderToolbar = () => {
+    const onCreateTask = () => {
+      const newTask = getElementById("newTask");
+      newTask.style.display = "block";
+      newTask.focus();
+    };
+
+    const onChangeLinkMode = () => setLinkMode((linkMode) => !linkMode);
```
Comment (justinline) [inline src/scripts/App.tsx:65]: issue: the variable is shadowed by the local state above. Maybe `state => !state` is preferable

File: src/scripts/Toolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    if (!props.tasksSelected) {
+      const { linkMode, onChangeLinkMode, onCreateTask } = props;
+      return (
+        <>
+          <button
+            aria-label="Change Link Mode"
+            onClick={onChangeLinkMode}
+            className={`Toolbar__button Toolbar__change-link-mode-button ${
+              linkMode ? "Toolbar__change-link-mode-button--active" : ""
+            } iconButton`}
```
Comment (justinline) [inline src/scripts/Toolbar.tsx:29]: thought: The css naming is a bit verbose - is it necessary? I feel like you could drop 1-2 words per classname. 
Also, shouldn't BEM be something like `Toolbar__button` and then `Toolbar__button-change-link-mode` 🤔 why did you put the `button` at the end of some and the start of others?

File: src/scripts/Toolbar.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+          />
+          <button
+            aria-label="Create Task"
```
Comment (justinline) [inline src/scripts/Toolbar.tsx:32]: praise: nice work putting on the labels ✨

File: src/scripts/useTasksSelected.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  useEffect(() => {
+    const onSelectionChanged = (event: Event) => {
+      const customEvent = event as CustomEvent<HTMLElement[]>;
+      setTasksSelected(customEvent.detail.length > 0);
+    };
+    const graph = getElementById("graph");
+    graph.addEventListener("selectionchanged", onSelectionChanged);
+    return () =>
+      graph.removeEventListener("selectionchanged", onSelectionChanged);
+  });
```
Comment (justinline) [inline src/scripts/useTasksSelected.ts:19]: question: this runs on every render without dependencies - is that intended? to run just once and set up the listener: `[]`

**Comment by justinline** [comment]
Finished 🏁

File: src/scripts/Toolbar.tsx (hunk 2 above)
Comment (etrobert) [reply src/scripts/Toolbar.tsx:29]: To answer your second point first:

From what I understand BEM has `multi-word-block__multi-word-element--multi-word-modifier`

But it does not specify the order of the words in `multi-word-element`

In this case the element is the 'Change Link Mode Button', in plain english

http://getbem.com/naming/

File: src/scripts/Toolbar.tsx (hunk 3 above)
Comment (etrobert) [reply src/scripts/Toolbar.tsx:32]: I was ready to press "Create Merge Request", then I though "Oh, Justin's going to clock me on that if I don't put them" 😂

File: src/scripts/useTasksSelected.ts (hunk 4 above)
Comment (etrobert) [reply src/scripts/useTasksSelected.ts:19]: Thanks for pointing it out 🤗 

I just had not thought of optimization yet

File: src/scripts/App.tsx (hunk 1 above)
Comment (etrobert) [reply src/scripts/App.tsx:65]: Well technically the variable _shadows_ the one above 😁 

I personally don't think shadowing should be avoided at all costs, especially in this case where we would loose readability by changing the name of the variable and the scope is so narrow that it's very clear what the function does 🤷

File: src/scripts/Toolbar.tsx (hunk 2 above)
Comment (etrobert) [reply src/scripts/Toolbar.tsx:29]: The reason that I did such verbose naming is for consistency.

To have a mapping 1 to 1 from the aria label to the event handlers to the css classes.

So clicking the button with `aria-label="Change Link Mode"` and class `Toolbar__change-link-mode-button` calls `onChangeLinkMode`

While it's not necessary, I find it very handy

Although the formatting was not great I just updated it to have one classname per line

Comment (justinline) [reply src/scripts/Toolbar.tsx]: regarding word order, it just makes search-ability via matching easier, but just a suggestion

File: src/scripts/App.tsx (hunk 1 above)
Comment (justinline) [reply src/scripts/App.tsx:65]: `setLinkMode(state => !state)` doesn't lose any readability in my mind. I see what I'm setting from the setter 🤔 

I like to personally avoid shadowing since it reduces effectiveness of selecting all occurrences and other productivity flows and don't see much benefit to the shadowing in this particular instance, but you're the author 😇

File: src/scripts/Toolbar.tsx (hunk 2 above)
Comment (justinline) [reply src/scripts/Toolbar.tsx]: i.e `Toolbar__button` search and replace would give me all related to this if the word order was `Toolbar__button-change-link-mode` - but currently `Toolbar__change-link-mode-button` would be missed

Comment (etrobert) [reply src/scripts/Toolbar.tsx]: That's a good point!! 🧐

I will implement it later, thanks 😊

File: src/scripts/App.tsx (hunk 1 above)
Comment (etrobert) [reply src/scripts/App.tsx:65]: That's also a great point 😄

You're on 🔥

Thanks for taking the time to share this thoughts 🤗
